File: OpenWA/sdk/java/src/main/java/com/rmyndharis/openwa/resources/StatusResource.java
package com.rmyndharis.openwa.resources;

import static com.rmyndharis.openwa.http.Http.encodeSegment;

import com.rmyndharis.openwa.OpenWAClient;
import com.rmyndharis.openwa.http.BinaryResponse;
import com.rmyndharis.openwa.http.HttpMethod;
import com.rmyndharis.openwa.model.SendImageStatusRequest;
import com.rmyndharis.openwa.model.SendTextStatusRequest;
import com.rmyndharis.openwa.model.SendVideoStatusRequest;
import com.rmyndharis.openwa.model.SendVoiceStatusRequest;
import com.rmyndharis.openwa.model.StatusListResult;
import com.rmyndharis.openwa.model.StatusResult;

/**
 * Status (Stories) resource — WhatsApp status updates.
 *
 * <p>NOTE: this is WhatsApp "Status/Stories", distinct from session lifecycle status.
 */
public final class StatusResource {
    private final OpenWAClient client;

    public StatusResource(OpenWAClient client) {
        this.client = client;
    }

    /** Get all status updates. */
    public StatusListResult list(String sessionId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/status",
            null,
            null,
            StatusListResult.class);
    }

    /** Get status updates from a specific contact. */
    public StatusListResult fromContact(String sessionId, String contactId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/status/" + encodeSegment(contactId),
            null,
            null,
            StatusListResult.class);
    }

    /** Fetch the stored media bytes for a status update (404 when there is no stored media). */
    public BinaryResponse media(String sessionId, String statusId) {
        return client.requestBytes(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/status/" + encodeSegment(statusId) + "/media",
            null);
    }

    /** Post a text status update. */
    public StatusResult sendText(String sessionId, SendTextStatusRequest body) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/status/send-text",
            null,
            body,
            StatusResult.class);
    }

    /** Post an image status update. */
    public StatusResult sendImage(String sessionId, SendImageStatusRequest body) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/status/send-image",
            null,
            body,
            StatusResult.class);
    }

    /** Post a video status update. */
    public StatusResult sendVideo(String sessionId, SendVideoStatusRequest body) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/status/send-video",
            null,
            body,
            StatusResult.class);
    }

    /**
     * Post an audio status as a voice note. WhatsApp plays one only as Ogg/Opus and neither engine
     * transcodes, so convert first with {@code MediaResource.convertVoice}. Requires an OPERATOR key.
     */
    public StatusResult sendVoice(String sessionId, SendVoiceStatusRequest body) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/status/send-voice",
            null,
            body,
            StatusResult.class);
    }

    /** Delete a status update by id. */
    public void delete(String sessionId, String statusId) {
        client.requestVoid(
            HttpMethod.DELETE,
            "/api/sessions/" + encodeSegment(sessionId) + "/status/" + encodeSegment(statusId),
            null,
            null);
    }
}

File: OpenWA/sdk/php/src/Resources/StatusResource.php
<?php

declare(strict_types=1);

namespace OpenWA\Resources;

use OpenWA\Http\HttpExecutor;

/**
 * Status (Stories) resource — WhatsApp status updates.
 *
 * Backed by src/modules/status/status.controller.ts.
 * NOTE: this is WhatsApp "Status/Stories", distinct from session lifecycle status.
 */
class StatusResource
{
    private HttpExecutor $http;

    public function __construct(HttpExecutor $http)
    {
        $this->http = $http;
    }

    /** @return array<string,mixed> */
    public function list(string $sessionId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status") ?? [];
    }

    /** @return array<string,mixed> */
    public function fromContact(string $sessionId, string $contactId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status/{$this->http->encodeSegment($contactId)}") ?? [];
    }

    /**
     * Fetch the stored media bytes for a status update (404 when there is no stored media).
     *
     * @return array{data: string, contentType: ?string}
     */
    public function media(string $sessionId, string $statusId): array
    {
        return $this->http->requestBinary('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status/{$this->http->encodeSegment($statusId)}/media");
    }

    /**
     * @param array<string,mixed> $body Body may include a `recipients` key (list<string> of JIDs).
     *                                   Required on the Baileys engine (absent/empty -> 400 there);
     *                                   ignored by whatsapp-web.js — omit it there.
     * @return array<string,mixed>
     */
    public function sendText(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status/send-text", [], $body);
    }

    /**
     * @param array<string,mixed> $body Body may include a `recipients` key (list<string> of JIDs).
     *                                   Required on the Baileys engine (absent/empty -> 400 there);
     *                                   ignored by whatsapp-web.js — omit it there.
     * @return array<string,mixed>
     */
    public function sendImage(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status/send-image", [], $body);
    }

    /**
     * @param array<string,mixed> $body Body may include a `recipients` key (list<string> of JIDs).
     *                                   Required on the Baileys engine (absent/empty -> 400 there);
     *                                   ignored by whatsapp-web.js — omit it there.
     * @return array<string,mixed>
     */
    public function sendVideo(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status/send-video", [], $body);
    }

    /**
     * Post an audio status as a voice note (OPERATOR). The body takes an `audio` wrapper and an
     * optional `recipients` list and an optional `backgroundColor` (#RRGGBB, Baileys only); it
     * carries no caption, since WhatsApp has nowhere to render one on a status voice note. WhatsApp plays one only as Ogg/Opus and neither engine transcodes —
     * convert first via MediaResource::convertVoice.
     *
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    public function sendVoice(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status/send-voice", [], $body);
    }

    public function delete(string $sessionId, string $statusId): void
    {
        $this->http->request('DELETE', "/api/sessions/{$this->http->encodeSegment($sessionId)}/status/{$this->http->encodeSegment($statusId)}");
    }
}
